### PR TITLE
Add a delete view

### PR DESCRIPTION
**Pull request by fnareoh**
Right now there is only a button delete on the user profile and not when you edit your profile which migth be a bit counter intuitive.
The position of the delete button is  a bit ugly.
The code forbids from deleting the account when the contest is on as mentioned in #95 .

### CODE REVIEW

**Comment by fnareoh** [comment]
"More red please!"
Sorry,I don't see what you mean ?

**Comment by seirl** [comment]
A few comments on the view itself:

- I'd like to have a list of things that will be deleted ("this will delete all your personal informations, your XXX, etc") in bold as a warning
- Would it be possible to ask the password for confirmation?

**Comment by zopieux** [comment]
@seirl [re: list of related stuff going to be deleted] this is a bit involved but it can be done by using [internal Django APIs](https://stackoverflow.com/a/26808663).

File: prologin/users/templates/users/edit.html [hunk 1]
```diff
@@ -24,4 +24,22 @@ <h1>{% blocktrans with edited_user.username as username %}Editing {{ username }}
...
+	<div class="Subhead Subhead--spacious">
+	  <h1 style = "color: #d9534f;"> <i class="fa fa-exclamation-triangle" aria-hidden="true"></i>
+ Delete account</h1>
+	</div>
+
+  
+  <p>{% trans "Are you sure you sure you want to delete your account ?" %}</p>
+  <p>{% trans "This will delete all your personnal informations, all your posts on the forum and all the code you have submited." %}</p>
+
+  <a href="{% url 'users:delete' user.id %}" class="btn btn-danger">
```
Comment (seirl) [inline prologin/users/templates/users/edit.html]: `edited_user.id`

File: prologin/users/templates/users/edit.html [hunk 2]
```diff
@@ -24,4 +24,22 @@ <h1>{% blocktrans with edited_user.username as username %}Editing {{ username }}
   {% crispy form %}
   <button class="btn btn-primary btn-block" type="submit"><i class="fa fa-save"></i> {% trans "Save" %}</button>
 </form>
+
+
+<br />
+<div class="row">
+<!-- Delete Account -->
+	<div class="Subhead Subhead--spacious">
+	  <h1 style = "color: #d9534f;"> <i class="fa fa-exclamation-triangle" aria-hidden="true"></i>
```
Comment (seirl) [inline prologin/users/templates/users/edit.html]: class + put this in .css

Comment (zopieux) [reply prologin/users/templates/users/edit.html]: the icon is too big for my taste

File: prologin/users/templates/users/edit.html [hunk 3]
```diff
@@ -24,4 +24,22 @@ <h1>{% blocktrans with edited_user.username as username %}Editing {{ username }}
...
+ Delete account</h1>
+	</div>
```
Comment (zopieux) [inline prologin/users/templates/users/edit.html]: whitespaces seem off

File: prologin/users/templates/users/edit.html [hunk 4]
```diff
@@ -24,4 +24,22 @@ <h1>{% blocktrans with edited_user.username as username %}Editing {{ username }}
...
+
+  
+  <p>{% trans "Are you sure you sure you want to delete your account ?" %}</p>
+  <p>{% trans "This will delete all your personnal informations, all your posts on the forum and all the code you have submited." %}</p>
```
Comment (zopieux) [inline prologin/users/templates/users/edit.html]: isn't “information” a plural singular?

File: prologin/users/views.py [hunk 5]
```diff
@@ -338,3 +338,47 @@ def render_to_response(self, context, **response_kwargs):
...
         return JsonResponse(results, safe=False)
+
+
+class DeleteView(PermissionRequiredMixin, View):
+    permission_required = 'users.delete'
+
+    def get(self, request, pk):
+        model = auth.get_user_model()
+        user = get_object_or_404(model, pk=pk)
+        return render(self.request, 'users/delete.html', {
```
Comment (zopieux) [inline prologin/users/views.py]: Just use a `FormView`?

File: prologin/users/views.py [hunk 6]
```diff
@@ -338,3 +338,47 @@ def render_to_response(self, context, **response_kwargs):
...
+            'delete_user': user,
+            'form': users.forms.PasswordConfirmForm
+        })
+
+    def post(self, request, pk):
+        model = auth.get_user_model()
+        user = get_object_or_404(model, pk=pk)
+
+        form = users.forms.PasswordConfirmForm(request.POST)
+        if form.is_valid():
```
Comment (zopieux) [inline prologin/users/views.py]: Yeah, definitely use a `FormView` :-)

File: prologin/users/views.py [hunk 7]
```diff
@@ -338,3 +338,47 @@ def render_to_response(self, context, **response_kwargs):
...
+            password = form.cleaned_data['password']
+            print(user.password)
```
Comment (zopieux) [inline prologin/users/views.py]: Thanks but no thanks

File: prologin/users/views.py [hunk 8]
```diff
@@ -338,3 +338,47 @@ def render_to_response(self, context, **response_kwargs):
...
+            if user.check_password(password):
+                if self.request.user.has_perm('users.edit-during-contest'):
+                    is_contest = False
+                else:
+                    is_contest = (self.request.current_edition.is_active
```
Comment (zopieux) [inline prologin/users/views.py]: This whole conditional should be computed in the `Edition` model class. Add it if is does not exist yet. Something like `qualification_finished(self)`

File: prologin/users/views.py [hunk 9]
```diff
@@ -338,3 +338,47 @@ def render_to_response(self, context, **response_kwargs):
...
+                        request,
+                        "You can't delete your account during the contest if you really want to, please send an email to info@prologin.fr"
+                    )
+                else:
+                    user.delete()
+                    messages.success(
+                        request, "Your account has been succesfuly deleted")
+                return render(self.request, 'users/delete_confirm.html', {})
+            else:
+                messages.error(request, "Please enter a correct password.")
```
Comment (zopieux) [inline prologin/users/views.py]: Use a proper form error, not a message. To do so, raise `ValidationError` in a [`clean`](https://docs.djangoproject.com/fr/2.0/ref/forms/validation/) method of the `PasswordConfirmForm`.

File: prologin/users/views.py [hunk 10]
```diff
@@ -338,3 +338,47 @@ def render_to_response(self, context, **response_kwargs):
...
+                    is_contest = (self.request.current_edition.is_active
+                                  and self.request.
+                                  current_events['qualification'].is_finished)
+                if is_contest:
+                    messages.error(
+                        request,
+                        "You can't delete your account during the contest if you really want to, please send an email to info@prologin.fr"
+                    )
+                else:
+                    user.delete()
```
Comment (zopieux) [inline prologin/users/views.py]: Put that in your `FormView` `form_valid(self)` handler. Add a call to `logout()` before deleting.

We'll later add a signal to display a staff alert somewhere.

**Review by zopieux — CHANGES_REQUESTED** [review]
tl;dr use a `FormView` instead

File: prologin/users/views.py (hunk 9 above)
Comment (fnareoh) [reply prologin/users/views.py]: How can I access the user in the `clean` method of the form to execute `user.check_password(password)` ? I have tried to defined an `__init__` method [inspired by this](https://stackoverflow.com/questions/5119994/get-current-user-in-django-form) but so far I can't make it work.

Comment (zopieux) [reply prologin/users/views.py]: You're right, it could work but I don't like putting stateful stuff in the form. While Googling I found [this](https://simpleisbetterthancomplex.com/tutorial/2016/08/15/how-to-create-a-password-confirmation-view.html). It's clean, does exactly what we want, uses modern Django stuff and even gives a nice decorator to add the mechanism to any view.

File: prologin/users/views.py (hunk 8 above)
Comment (fnareoh) [reply prologin/users/views.py]: Done, I have also replaced it in the `EditUserView`.

Comment (zopieux) [reply prologin/users/views.py]: Neat!

**Comment by zopieux** [comment]
For super users deleting others, please add a "recopy username to delete here" field. That mean you could do a generic confirm page (view) taking any Form class as a class attribute, one for users deleting self, one for SU. Two views (and two forms, one for each) both inheriting a third, generic Confirm view.

File: prologin/users/views.py [hunk 11]
```diff
@@ -338,3 +334,59 @@ def render_to_response(self, context, **response_kwargs):
...
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        model = auth.get_user_model()
+        context['delete_user'] = get_object_or_404(model, pk=self.kwargs['pk'])
+        context['form'] = users.forms.ConfirmDeleteForm(self.request.user)
+        return context
+
+    def get_form_kwargs(self):
+        kwargs = super().get_form_kwargs()
+        kwargs['user'] = self.request.user
```
Comment (seirl) [inline prologin/users/views.py:354]: Do you really need this? It's already `{{ request.user }}`

File: prologin/users/views.py [hunk 12]
```diff
@@ -338,3 +334,59 @@ def render_to_response(self, context, **response_kwargs):
...
+                context['form'] = form
+                return render(self.request, 'users/delete.html', context)
+        except KeyError:
+            pass
+
+        is_contest = contest.models.Edition.qualification_finished(self)
+        if is_contest:
+            messages.error(
+                self.request,
+                "You can't delete your account during the contest if you really want to, please send an email to info@prologin.fr"
```
Comment (seirl) [inline prologin/users/views.py]: Two sentences

File: prologin/users/views.py [hunk 13]
```diff
@@ -338,3 +334,59 @@ def render_to_response(self, context, **response_kwargs):
...
+            )
+        else:
+            delete_username = delete_user.username
+            delete_user.delete()
+            messages.success(self.request,
+                             delete_username + " account has been succesfuly deleted")
```
Comment (seirl) [inline prologin/users/views.py]: successfully

File: prologin/contest/models.py [hunk 14]
```diff
@@ -98,6 +98,15 @@ def phase(self):
     def __str__(self):
         return "Prologin {}".format(self.year)
 
+    def qualification_finished(self):
```
Comment (seirl) [inline prologin/contest/models.py:101]: This is really weird. The function is called "qualification_finished" but what it actually checks is whether the current user has the permission to edit its profile. Plus, this has nothing to do in the Edition model.

File: prologin/users/forms.py [hunk 15]
```diff
@@ -14,6 +14,7 @@
 User = get_user_model()
 
 
+
```
Comment (seirl) [inline prologin/users/forms.py]: Remove

File: prologin/users/forms.py [hunk 16]
```diff
@@ -155,3 +156,16 @@ def __init__(self, *args, **kwargs):
...
+class ConfirmDeleteForm(forms.ModelForm):
+    confirm_password = forms.CharField(widget=forms.PasswordInput())
+
+    class Meta:
+        model = User
+        fields = ('confirm_password', )
+
+    def __init__(self, user, **kwargs):
+        super().__init__(**kwargs)
+        if user.is_staff:
```
Comment (seirl) [inline prologin/users/forms.py:169]: `if user != self.request.user` would probably be more appropriate here?

File: prologin/users/templates/users/confirm_password.html [hunk 17]
```diff
@@ -0,0 +1,9 @@
+{% extends 'prologin/base.html' %}
+
+{% block content %}
+  <form method="post">
+    {% csrf_token %}
+    {{ form.as_p }}
+    <button type="submit">Confirm credentials</button>
+  </form>
+{% endblock %}
```
Comment (seirl) [inline prologin/users/templates/users/confirm_password.html]: Newline at end of file

File: prologin/users/templates/users/confirm_password.html [hunk 18]
```diff
@@ -0,0 +1,9 @@
+{% extends 'prologin/base.html' %}
```
Comment (seirl) [inline prologin/users/templates/users/confirm_password.html]: Shouldn't that file be removed?

File: prologin/users/views.py (hunk 13 above)
Comment (seirl) [reply prologin/users/views.py]: and `'s account`

File: prologin/users/views.py [hunk 19]
```diff
@@ -338,3 +334,59 @@ def render_to_response(self, context, **response_kwargs):
                 self.request),
         } for user in self.get_queryset()]
         return JsonResponse(results, safe=False)
+
+
+class DeleteView(PermissionRequiredMixin, UpdateView):
```
Comment (seirl) [inline prologin/users/views.py:339]: Do we really want an UpdateView here? We're not updating anything. I fear it might silently update some fields of the user.

File: prologin/users/views.py [hunk 20]
```diff
@@ -338,3 +334,59 @@ def render_to_response(self, context, **response_kwargs):
...
+        model = auth.get_user_model()
+        user = self.request.user
+        delete_user = get_object_or_404(model, pk=self.kwargs['pk'])
+        password = form.cleaned_data['confirm_password']
+
+        if not user.check_password(password):
+            form.add_error('confirm_password', 'Password does not match.')
+            context = self.get_context_data()
+            context['form'] = form
+            return render(self.request, 'users/delete.html', context)
```
Comment (seirl) [inline prologin/users/views.py]: I really think you shouldn't do that by hand, the form should be able to do that by itself.

File: prologin/users/views.py [hunk 21]
```diff
@@ -338,3 +334,59 @@ def render_to_response(self, context, **response_kwargs):
...
+
+        try:
+            username = form.cleaned_data['confirm_user']
+            if username != delete_user.username:
+                form.add_error('confirm_user', 'Username does not match.')
+                context = self.get_context_data()
+                context['form'] = form
+                return render(self.request, 'users/delete.html', context)
+        except KeyError:
```
Comment (seirl) [inline prologin/users/views.py]: What is this supposed to catch? Only use try: except: around the relevant parts

**Review by seirl — CHANGES_REQUESTED** [review]

File: prologin/users/views.py (hunk 11 above)
Comment (fnareoh) [reply prologin/users/views.py:354]: I need it to init the form and have a user_confirm field field only if the user is_staff

File: prologin/contest/models.py (hunk 14 above)
Comment (fnareoh) [reply prologin/contest/models.py:101]: [zopieux told me to do so](https://github.com/prologin/site/pull/98/files/b497894d051e123152deca5de5592a2c367014b3#diff-179adfdfcc0692485b7637aa929677c2)

File: prologin/users/views.py (hunk 21 above)
Comment (fnareoh) [reply prologin/users/views.py]: It catch's the KeyError when there is no field `confirm_user` when it's only the user deleting it's own account.

File: prologin/users/views.py (hunk 19 above)
Comment (fnareoh) [reply prologin/users/views.py:339]: It's a remaining of [this code recommendation](https://simpleisbetterthancomplex.com/tutorial/2016/08/15/how-to-create-a-password-confirmation-view.html). It already had me worried but i couldn't change it, else it failed. I'll see if it's better now.

File: prologin/users/templates/users/confirm_password.html (hunk 18 above)
Comment (fnareoh) [reply prologin/users/templates/users/confirm_password.html]: yes

File: prologin/users/views.py (hunk 20 above)
Comment (fnareoh) [reply prologin/users/views.py]: I agree, but I didn't found out how.

File: prologin/contest/models.py (hunk 14 above)
Comment (seirl) [reply prologin/contest/models.py:101]: He probably was talking about this line only:

    self.request.current_edition.is_active and self.request.current_events['qualification'].is_finished

File: prologin/users/views.py (hunk 19 above)
Comment (fnareoh) [reply prologin/users/views.py:339]: Getting it to FormView causes (I don't know why) problems of permissions, a regular user can't access the delete page of it's own count it reloads indefinitely a login page

File: prologin/users/templates/users/edit.html [hunk 22]
```diff
@@ -24,4 +24,22 @@ <h1>{% blocktrans with edited_user.username as username %}Editing {{ username }}
   {% crispy form %}
   <button class="btn btn-primary btn-block" type="submit"><i class="fa fa-save"></i> {% trans "Save" %}</button>
 </form>
+
+
+<br />
+<div class="row">
+<!-- Delete Account -->
+	<div class="Subhead Subhead--spacious">
```
Comment (seirl) [inline prologin/users/templates/users/edit.html]: lowercase class names

File: prologin/users/views.py [hunk 23]
```diff
@@ -382,11 +382,12 @@ def form_valid(self, form, *args, **kwargs):
         if is_contest:
             messages.error(
                 self.request,
-                "You can't delete your account during the contest if you really want to, please send an email to info@prologin.fr"
+                "You can't delete your account during the contest. "
```
Comment (zopieux) [inline prologin/users/views.py]: `_()` (translation, add final dot)

File: prologin/users/views.py [hunk 24]
```diff
@@ -382,11 +382,12 @@ def form_valid(self, form, *args, **kwargs):
...
+                "If you really want to do so, please send an email to info@prologin.fr"
             )
         else:
             delete_username = delete_user.username
             delete_user.delete()
             messages.success(self.request,
-                             delete_username + " account has been succesfuly deleted")
+                             delete_username + "'s account has been succesfully deleted")
```
Comment (zopieux) [inline prologin/users/views.py]: `_("%(user)s's account has been successfully deleted.") % {'user': deleted_username}` (translation, typo, final dot)

File: prologin/users/templates/users/edit.html [hunk 25]
```diff
@@ -24,4 +24,21 @@ <h1>{% blocktrans with edited_user.username as username %}Editing {{ username }}
...
+	  <h1 style = "color: #d9534f;"> <i class="fa fa-exclamation-triangle" aria-hidden="true"></i> Delete account</h1>
+	</div>
+
+  
+  <p>{% trans "Are you sure you sure you want to delete your account ?" %}</p>
```
Comment (zopieux) [inline prologin/users/templates/users/edit.html]: **you sure you sure*, no space before *?*

File: prologin/users/templates/users/edit.html [hunk 26]
```diff
@@ -24,4 +24,21 @@ <h1>{% blocktrans with edited_user.username as username %}Editing {{ username }}
...
+  <p>{% trans "This will delete all your personnal information, all your posts on the forum and all the code you have submited." %}</p>
```
Comment (zopieux) [inline prologin/users/templates/users/edit.html]: *personal*, *submitted*

File: prologin/users/templates/users/edit.html [hunk 27]
```diff
@@ -24,4 +24,21 @@ <h1>{% blocktrans with edited_user.username as username %}Editing {{ username }}
...
+
+  <a href="{% url 'users:delete' edited_user.id %}" class="btn btn-danger">
+  {% trans "Delete Account" %}</a>
```
Comment (zopieux) [inline prologin/users/templates/users/edit.html]: No capital *A*

File: prologin/users/views.py [hunk 28]
```diff
@@ -338,3 +334,60 @@ def render_to_response(self, context, **response_kwargs):
...
+
+    def form_valid(self, form, *args, **kwargs):
+
+        model = auth.get_user_model()
+        user = self.request.user
+        delete_user = get_object_or_404(model, pk=self.kwargs['pk'])
+        password = form.cleaned_data['confirm_password']
+
+        if not user.check_password(password):
+            form.add_error('confirm_password', 'Password does not match.')
```
Comment (zopieux) [inline prologin/users/views.py]: Translation. Also there is just one password so just use eg. *Wrong password.* instead.

File: prologin/users/views.py [hunk 29]
```diff
@@ -338,3 +334,60 @@ def render_to_response(self, context, **response_kwargs):
...
+    template_name = 'users/delete.html'
+    form_class = users.forms.ConfirmDeleteForm
+    context_object_name = 'delete_user'
+    model = auth.get_user_model()
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        model = auth.get_user_model()
+        context['delete_user'] = get_object_or_404(model, pk=self.kwargs['pk'])
+        context['form'] = users.forms.ConfirmDeleteForm(self.request.user)
```
Comment (zopieux) [inline prologin/users/views.py]: Nah. You're doing the work of the `FormView` here.

File: prologin/users/views.py [hunk 30]
```diff
@@ -338,3 +334,60 @@ def render_to_response(self, context, **response_kwargs):
...
+        if not user.check_password(password):
+            form.add_error('confirm_password', 'Password does not match.')
+            context = self.get_context_data()
+            context['form'] = form
+            return render(self.request, 'users/delete.html', context)
+
+        try:
+            username = form.cleaned_data['confirm_user']
+            if username != delete_user.username:
+                form.add_error('confirm_user', 'Username does not match.')
```
Comment (zopieux) [inline prologin/users/views.py]: cf. password comment

**Review by zopieux — CHANGES_REQUESTED** [review]

File: prologin/users/views.py (hunk 28 above)
Comment (fnareoh) [reply prologin/users/views.py]: I've look at the other error on form, what signifies the translation ?

**Comment by zopieux** [comment]
Reworked and merged in 96c97a5c8f91f23766def270a329dd158abeb289. Thanks!
